### PR TITLE
:sparkles: Instant ID

### PR DESCRIPTION
**Pull request by huchenlei**
## Instant ID project
https://github.com/InstantID/InstantID

Instant ID uses a combination of ControlNet and IP-Adapter to control the facial features in the diffusion process. One unique design for Instant ID is that it passes facial embedding from IP-Adapter projection as crossattn input to the ControlNet unet. Normally the crossattn input to the ControlNet unet is prompt's text embedding.
image

## Download models
You need to download following models and put them under `{A1111_root}/models/ControlNet` directory. It is also required to rename models to `ip-adapter_instant_id_sdxl` and `control_instant_id_sdxl` so that they can be correctly recognized by the extension.
- [ipadapter model](https://huggingface.co/InstantX/InstantID/resolve/main/ip-adapter.bin?download=true)
- [ControlNet model](https://huggingface.co/InstantX/InstantID/resolve/main/ControlNetModel/diffusion_pytorch_model.safetensors?download=true)

## How to use
InstantID takes 2 models on the UI. You should always set the ipadapter model as first model, as the ControlNet model takes the output from the ipadapter model. (ipadapter model should be hooked first)

### Unit 0 Setting
You must set ip-adapter unit right before the ControlNet unit. The projected face embedding output of IP-Adapter unit will be used as part of input to the next ControlNet unit.

### Unit 1 Setting
The ControlNet unit accepts a keypoint map of 5 facial keypoints. You are not restricted to use the facial keypoints of the same person you used in Unit 0. Here I use a different person's facial keypoints.

## CFG
It is recommended to set CFG 4~5 to get best result. Depending on sampling method and base model this number may vary, but generally you need to use CFG scale a little bit less than normal CFG.
## Follow-up work
- Make sd-webui-openpose-editor able to edit the facial keypoints in preprocessor result preview.
- Currently even if you are using the same face for both model, the insightface preprocessor will run twice. We need to find a way to cache the result and only run the model once.
- Support multiple face inputs.

## Note
As the insightface's github release currently do not have antelopev2 model, we are downloading from a huggingface mirror https://huggingface.co/DIAMONIK7777/antelopev2. If you are in mainland China and don't have good internet connection to huggingface, you can manually download the model from somewhere else and place them under `extensions/sd-webui-controlnet/annotators/downloads/insightface/models/antelopev2`.

### CODE REVIEW

**Comment by liubo0902** [comment]
with cfg, the weight of the ip-adapter should be low, as in the demo code of instand_id, the weight is 0.5, you can try this

File: scripts/processor.py [hunk 1]
```diff
@@ -736,8 +740,49 @@ def run_model(self, imgs: Union[Tuple[np.ndarray], np.ndarray], **kwargs):
...
+        self.load_model()
+        face_info = self.model.get(img)
+        if not face_info:
+            raise Exception(f"Insightface: No face found in image.")
+        if len(face_info) > 1:
+            logger.warn("Insightface: More than one face is detected in the image. "
+                        f"Only the first one will be used.")
+        # only use the maximum face
+        face_info = sorted(face_info, key=lambda x:(x['bbox'][2]-x['bbox'][0])*x['bbox'][3]-x['bbox'][1])[-1]
+        return RawInstantIdInput(draw_kps(img, face_info['kps']), face_info['embedding']), False
```
Comment (huchenlei) [inline scripts/processor.py:781]: Step1: Accepting raw inputs, returns faceid processed results.

File: scripts/controlnet.py [hunk 2]
```diff
@@ -957,6 +954,12 @@ def store_detected_map(detected_map, module: str) -> None:
             if is_image:
                 control, detected_map = Script.detectmap_proc(detected_map, unit.module, resize_mode, h, w)
                 store_detected_map(detected_map, unit.module)
+            elif control_model_type == ControlModelType.InstantID:
+                assert isinstance(detected_map, tuple)
+                raw_input = detected_map
+                resized_keypoints, detected_map = Script.detectmap_proc(raw_input.keypoints, unit.module, resize_mode, h, w)
```
Comment (huchenlei) [inline scripts/controlnet.py:960]: Step3: Keypoints map get resized to generation target's height/width.

File: scripts/controlmodel_ipadapter.py [hunk 3]
```diff
@@ -371,6 +387,22 @@ def get_image_embeds_faceid(self, insightface_outputs: List[torch.Tensor]):
         uncond_image_prompt_embeds = uncond_image_prompt_embeds.reshape(batch_size, -1, c)
         return image_prompt_embeds, uncond_image_prompt_embeds
 
+    @torch.inference_mode()
+    def get_image_embeds_instantid(self, prompt_image_emb):
```
Comment (huchenlei) [inline scripts/controlmodel_ipadapter.py:391]: Step2: Calculate projected face embedding.

File: scripts/controlnet.py [hunk 4]
```diff
@@ -1099,6 +1097,24 @@ def recolor_intensity_post_processing(x):
...
+            if param.control_model_type == ControlModelType.InstantID:
+                # For instant_id we always expect ip-adapter model followed
+                # by ControlNet model.
+                assert i > 0, "InstantID control model should follow ipadapter model."
+                ip_adapter_param = forward_params[i - 1]
+                assert ip_adapter_param.control_model_type == ControlModelType.IPAdapter, \
+                        "InstantID control model should follow ipadapter model."
+                control_model = ip_adapter_param.control_model
+                assert hasattr(control_model, "image_emb")
+                param.hint_cond = InstantIdInput(
```
Comment (huchenlei) [inline scripts/controlnet.py:1109]: Step4: Pass projected face embedding to ControlNet.

File: scripts/hook.py [hunk 5]
```diff
@@ -559,6 +563,11 @@ def forward(self, x, timesteps=None, context=None, y=None, **kwargs):
                 assert param.used_hint_cond is not None, f"Controlnet is enabled but no input image is given"
 
                 hint = param.used_hint_cond
+                # Unpack inputs for InstantID.
```
Comment (huchenlei) [inline scripts/hook.py:566]: Step5: Set control image(`hint`) and crossattn cond(`context`) for ControlNet.

File: scripts/controlmodel_ipadapter.py [hunk 6]
```diff
@@ -371,6 +387,22 @@ def get_image_embeds_faceid(self, insightface_outputs: List[torch.Tensor]):
...
+        else:
+            prompt_image_emb = torch.tensor(prompt_image_emb)
+
+        prompt_image_emb = prompt_image_emb.to(device=self.device, dtype=torch.float32)
+        prompt_image_emb = prompt_image_emb.reshape([1, -1, image_proj_model_in_features])
+        return (
+            self.image_proj_model(prompt_image_emb),
+            self.image_proj_model(torch.zeros_like(prompt_image_emb)),
+        )
+
```
Comment (huchenlei) [inline scripts/controlmodel_ipadapter.py:405]: Step2: Calculate projected face embedding with ipadapter weights.

File: scripts/processor.py (hunk 1 above)
Comment (huchenlei) [inline scripts/processor.py:781]: Step1: Accepting raw inputs, returns faceid processed results.

File: scripts/controlnet.py [hunk 7]
```diff
@@ -957,6 +954,12 @@ def store_detected_map(detected_map, module: str) -> None:
             if is_image:
                 control, detected_map = Script.detectmap_proc(detected_map, unit.module, resize_mode, h, w)
                 store_detected_map(detected_map, unit.module)
+            elif control_model_type == ControlModelType.InstantID:
+                assert isinstance(detected_map, tuple)
+                raw_input = detected_map
+                resized_keypoints, detected_map = Script.detectmap_proc(raw_input.keypoints, unit.module, resize_mode, h, w)
+                control = ResizedInstantIdInput(resized_keypoints, raw_input.embedding)
+                store_detected_map(detected_map, unit.module)
```
Comment (huchenlei) [inline scripts/controlnet.py:962]: Step3: Keypoints map get resized to generation target's height/width.

File: scripts/controlnet.py [hunk 8]
```diff
@@ -1099,6 +1097,24 @@ def recolor_intensity_post_processing(x):
...
+                assert i > 0, "InstantID control model should follow ipadapter model."
+                ip_adapter_param = forward_params[i - 1]
+                assert ip_adapter_param.control_model_type == ControlModelType.IPAdapter, \
+                        "InstantID control model should follow ipadapter model."
+                control_model = ip_adapter_param.control_model
+                assert hasattr(control_model, "image_emb")
+                param.hint_cond = InstantIdInput(
+                    param.hint_cond.resized_keypoints,
+                    control_model.image_emb,
+                )
```
Comment (huchenlei) [inline scripts/controlnet.py:1112]: Step4: Pass projected face embedding to ControlNet.

File: scripts/hook.py [hunk 9]
```diff
@@ -559,6 +563,11 @@ def forward(self, x, timesteps=None, context=None, y=None, **kwargs):
...
+                if param.control_model_type == ControlModelType.InstantID:
+                    assert isinstance(hint, InstantIdInput)
+                    context = hint.projected_embedding.eval(cond_mark).to(x.device, dtype=x.dtype)
+                    hint = hint.resized_keypoints.to(x.device, dtype=x.dtype)
```
Comment (huchenlei) [inline scripts/hook.py:570]: Step5: Set control image(hint) and crossattn cond(context) for ControlNet.

**Comment by erhan-** [comment]
Testing this.

First run:

```
        raise RuntimeError("Failed downloading url %s" % url)
    RuntimeError: Failed downloading url https://github.com/deepinsight/insightface/releases/download/v0.7/antelopev2.zip
```

https://github.com/deepinsight/insightface/issues/1896#issuecomment-1023867304

**Comment by hablaba** [comment]
This is awesome! I’ve played with InstantID in diffusers and it has a lot of potential. 

I’m also testing out your PR. I did notice the examples in InstantID results use super low CFG at 3.5. Higher and I get pretty bad results. But yes something seems a bit off in the implementation because I don’t get great likeness. 

Two other notes - I think there needs to be a way to adjust control net scale and ip adapter scale separately. Often times you need to tweak them independently. Another really cool opportunity would be to allow a secondary pose image. If it’s not included the pose image would default to the face image. That would let you generate with different angles/poses. 

Great work! Appreciate you working on this so quick. Happy to help beta test some more

**Comment by beansfotos** [comment]
Does this work? Will it be included in a future update or do we have to manually install?

**Comment by huchenlei** [comment]
> This is awesome! I’ve played with InstantID in diffusers and it has a lot of potential.
> 
> I’m also testing out your PR. I did notice the examples in InstantID results use super low CFG at 3.5. Higher and I get pretty bad results. But yes something seems a bit off in the implementation because I don’t get great likeness.
> 
> Two other notes - I think there needs to be a way to adjust control net scale and ip adapter scale separately. Often times you need to tweak them independently. Another really cool opportunity would be to allow a secondary pose image. If it’s not included the pose image would default to the face image. That would let you generate with different angles/poses.
> 
> Great work! Appreciate you working on this so quick. Happy to help beta test some more

Thanks for your testing! I have made insightface into 2 separate units. Now you can adjust weight for each model and optionally you can pass a custom facial landmark now.

**Comment by hablaba** [comment]
> Thanks for your testing! I have made insightface into 2 separate units. Now you can adjust weight for each model and optionally you can pass a custom facial landmark now.

I tested your changes and it seems to be working great, including with different face image and pose image!

Really appreciate your work here and I’m a bit shocked at how fast you got it implemented. 

The low CFG requirement is still a odd quirk but that seems to really be an InstantID issue. Maybe they’ll fix it in a future model

**Comment by aminesoulaymani** [comment]
Amazing, I spent hours trying to deal with OOM's using comfyui even on 768*768px with poor results (I5, Geforce 3060 6Gb Vram), after the sd-webui-controlnet update, automatic111 is able to produce 1024*1024 near-perfect results, no OOM's at all, no "Low vram" checked, everything smooth and nice. The results are better than any Lora I trained for days, about 3 minutes with your extension, you MVP

**Comment by Kvisaz** [comment]
It create face, but totally destroy all styles. Every time - gradients, abstract patterns, weird style

It seems to be totally broken realisation 

What i want in style:



What i get



````
nice girl,comic books illustration,
Steps: 20, Sampler: DPM++ 2M Karras, CFG scale: 3, Seed: 1000, Size: 512x512, Model hash: 676f0d60c8, Model: dreamshaperXL_turboDpmppSDE, ControlNet 0: "Module: instant_id_face_embedding, Model: ip-adapter_instant_id_sdxl [eb2d3ec0], Weight: 1, Resize Mode: Crop and Resize, Low Vram: False, Processor Res: 512, Guidance Start: 0, Guidance End: 1, Pixel Perfect: False, Control Mode: Balanced, Hr Option: Both, Save Detected Map: True", ControlNet 1: "Module: instant_id_face_keypoints, Model: control_instant_id_sdxl [c5c25a50], Weight: 1, Resize Mode: Crop and Resize, Low Vram: False, Processor Res: 512, Guidance Start: 0, Guidance End: 1, Pixel Perfect: False, Control Mode: Balanced, Hr Option: Both, Save Detected Map: True", Version: v1.7.0
```

And this weird style every time, every generation.

Dont use this plugin, it doesnt work.

**Pull request merged**